coll/tuned: Change the default collective algorithm selection

wckzhang opened the pull request:
The default algorithm selections were out of date and not performing
well. After gathering data from OMPI developers, new default algorithm
decisions were selected for:

    allgather
    allgatherv
    allreduce
    alltoall
    alltoallv
    barrier
    bcast
    gather
    reduce
    reduce_scatter_block
    reduce_scatter
    scatter

Signed-off-by: William Zhang <wilzhang@amazon.com>
(cherry picked from commit ce40cfbaa53406be71319041e13e893b0def7ad9)

wckzhang commented on ompi/mca/coll/tuned/coll_tuned_decision_fixed.c:
```diff
@@ -48,48 +50,199 @@ ompi_coll_tuned_allreduce_intra_dec_fixed(const void *sbuf, void *rbuf, int coun
     OPAL_OUTPUT((ompi_coll_tuned_stream, "ompi_coll_tuned_allreduce_intra_dec_fixed"));
 
-    /**
-     * Decision function based on MX results from the Grig cluster at UTK.
```
Done

wzamazon commented on ompi/mca/coll/tuned/coll_tuned_decision_fixed.c:
```diff
@@ -357,108 +701,158 @@ int ompi_coll_tuned_reduce_intra_dec_fixed( const void *sendbuf, void *recvbuf,
+                alg = 6;
+            } else {
+                alg = 5;
+            }
```
Why is not there a section for ```communicator_size < 1024```?

wckzhang replied:
See above comment

wzamazon replied:
Thanks.

wckzhang commented on ompi/mca/coll/tuned/coll_tuned_decision_fixed.c:
```diff
@@ -48,48 +58,199 @@ ompi_coll_tuned_allreduce_intra_dec_fixed(const void *sbuf, void *rbuf, int coun
+                alg = 6;
+            }
+        } else if (communicator_size < 8) {
+            if (total_dsize < 15) {
```
Huh year that's weird, will change

wckzhang commented on ompi/mca/coll/tuned/coll_tuned_decision_fixed.c:
```diff
@@ -48,48 +58,199 @@ ompi_coll_tuned_allreduce_intra_dec_fixed(const void *sbuf, void *rbuf, int coun
+                alg = 2;
+            } else {
+                alg = 6;
+            }
```
Since the optimal algorithms are the same for communicator size 512 and 1024 it's redundant having another section

wzamazon replied:
I see. thanks.

jsquyres commented:
These changes need to go in on master before they can go in on the v4.1.x branch.

wckzhang commented:
> These changes need to go in on master before they can go in on the v4.1.x branch.

The master branch changes haven't been merged either so I've been pushing fixes in to both PR's at the same time

jsquyres commented:
> The master branch changes haven't been merged either so I've been pushing fixes in to both PR's at the same time

Cool.  I just converted this PR to "draft" -- let's leave it as draft until the corresponding master PR has been merged and this PR is therefore ready for v4.1.x.

Thanks!

wckzhang commented:
> > The master branch changes haven't been merged either so I've been pushing fixes in to both PR's at the same time
> 
> Cool. I just converted this PR to "draft" -- let's leave it as draft until the corresponding master PR has been merged and this PR is therefore ready for v4.1.x.
> 
> Thanks!

Thanks Jeff, PR merged and I updated cherry pick ID.

jsquyres commented:
Here's the corresponding master PR: #7730
